internal/stanza/converter.go

```diff
@@ -582,3 +571,66 @@ var sevTextMap = map[entry.Severity]string{
 	entry.Fatal3:  "Fatal3",
 	entry.Fatal4:  "Fatal4",
 }
+
+// pairSep is chosen to be an invalid byte for a utf-8 sequence
+// making it more unlikely to be hit
+var pairSep = []byte{0xfe}
+
+// emptyResourceID is the ID returned by getResourceID when it is passed an empty resource.
+// This specific number is chosen as it is the starting offset of fnv64.
+const emptyResourceID uint64 = 14695981039346656037
+
+func getResourceID(resource map[string]string) uint64 {
+	if len(resource) == 0 {
+		return emptyResourceID
+	}
+
+	var fnvHash = fnv.New64a()
+	var fnvHashOut = make([]byte, 0, 16)
+	var keySlice = make([]string, 0, len(resource))
+	var escapedSlice = make([]byte, 0, 64)
+
+	for k := range resource {
+		keySlice = append(keySlice, k)
+	}
+
+	// In order for this to be deterministic, we need to sort the map. Using range, like above,
+	// has no guarantee about order.
+	sort.Strings(keySlice)
+	for _, k := range keySlice {
+		escapedSlice = appendEscapedPairSeparator(escapedSlice[:0], k)
+		fnvHash.Write(escapedSlice)
+		fnvHash.Write(pairSep)
+
+		escapedSlice = appendEscapedPairSeparator(escapedSlice[:0], resource[k])
+		fnvHash.Write(escapedSlice)
+		fnvHash.Write(pairSep)
+	}
+
+	fnvHashOut = fnvHash.Sum(fnvHashOut)
+	return binary.BigEndian.Uint64(fnvHashOut)
+}
+
+// appendEscapedPairSeparator escapes (prefixes) "pair_sep" and byte '0xff' with byte '0xff', and appends it to the
+// incoming buffer. It returns the appended buffer.
+func appendEscapedPairSeparator(buf []byte, s string) []byte {
+	const escapeByte byte = '\xff'
+
+	if len(s) > cap(buf) {
+		newBuf := make([]byte, 2*len(s))
+		copy(newBuf, buf)
+		buf = newBuf
+	}
+
+	sBytes := []byte(s)
+	for _, b := range sBytes {
+		switch b {
+		case escapeByte, pairSep[0]:
+			buf = append(buf, escapeByte)
+		}
+
+		buf = append(buf, b)
+	}
+
+	return buf
+}
```

Comment:
I can appreciate the theoretical need for this, but I wonder how necessary it really is in practice. We're basically making a tradeoff between correctness and performance.

The performance hit seems potentially substantial, because every single byte of every single log event's resource attribute keys and values are reprocessed here.

The correctness benefit seems likely to be non-zero, but nearly zero, as we don't really expect to see these extremely unusual bytes in the resource attributes.

I hesitate to allow the extremely rare case, but if I'm understanding it correctly, it does seem quite far fetched. What are your thoughts on potentially skipping this step? (Also it would be nice to see with/without benchmarks so we can characterize the benefit)

Reply:
Yeah, I definitely see what you're saying here. I also think it's extremely unlikely for this to actually be necessary in 99.9% of cases. To give a specific example of something that would be a problem:

If you have a resource:
```
{
  "AB\xfe": "DF"
}
```

This, without escaping, would be ambiguous with
```
{
  "AB": "\xfeDF"
}
```

Meaning that one resource would overwrite the other when aggregating. 

This is the performance difference:

With escaping:
```
goos: darwin
goarch: amd64
pkg: github.com/open-telemetry/opentelemetry-collector-contrib/internal/stanza
cpu: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
BenchmarkGetResourceID-12    	 1221342	       997.3 ns/op	     152 B/op	       2 allocs/op
```

Without escaping:
```
goos: darwin
goarch: amd64
pkg: github.com/open-telemetry/opentelemetry-collector-contrib/internal/stanza
cpu: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
BenchmarkGetResourceID-12    	 1894089	       622.4 ns/op	     152 B/op	       2 allocs/op
PASS
```

That's gonna shave about ~300 ns per entry. Which isn't completely insignificant.
Heres the converter benchmark after removing:
```
goos: darwin
goarch: amd64
pkg: github.com/open-telemetry/opentelemetry-collector-contrib/internal/stanza
cpu: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
BenchmarkConverter/worker_count=1-12         	       1	6252154042 ns/op	3061890344 B/op	61018225 allocs/op
BenchmarkConverter/worker_count=2-12         	       1	4004045147 ns/op	3061885672 B/op	61018174 allocs/op
BenchmarkConverter/worker_count=4-12         	       1	3250401093 ns/op	3061871160 B/op	61018133 allocs/op
BenchmarkConverter/worker_count=6-12         	       1	3000911347 ns/op	3061841568 B/op	61018143 allocs/op
BenchmarkConverter/worker_count=8-12         	       1	3004432356 ns/op	3061879752 B/op	61018132 allocs/op
```

So yeah, it helps to remove that step. I don't know whether sacrificing correctness is the call to make here or not. I don't think it's common at all. But also I can see it being a real pain to debug if/when an issue with it shows up in the real world.

My opinion is to keep it for correctness sake, unless we have some guarantee that the input data here is utf-8.

Reply:
In general, resource identification is meant to be human readable, so it seems these particular delimiters should be sufficiently obscure so as not to be much concern. Additionally, resource keys _should_ follow the [resource semantic conventions](https://github.com/open-telemetry/opentelemetry-specification/tree/main/specification/resource/semantic_conventions), which will never prescribe keys with those bytes. 

I feel pretty strongly we should remove this. It's adding 50% to the cost of hashing the resource ID for the benefit of a tremendously unlikely scenario that we do not need to accommodate.

Reply:
Alright, fair enough, if we have a strong expectation for these strings to be human readable, we're probably fine removing it. I'll remove the escaping, then.